fpga_interchange/site_router.cc

```diff
@@ -39,669 +49,597 @@ void SiteRouter::unbindBel(CellInfo *cell)
+                         user.cell->type.c_str(ctx));
+            }
+        }
     }
 
-    void free_node(Node node) { node_storage->free_node(nodes, node); }
+    log_info(" Consumed wires:\n");
+    for (auto &wire_pair : site_arch->wire_to_nets) {
+        const SiteWire &site_wire = wire_pair.first;
+        if (site_wire.type != SiteWire::SITE_WIRE) {
+            continue;
+        }
+        WireId wire = site_wire.wire;
+        const NetInfo *net = wire_pair.second.net->net;
+        log_info("  - %s is bound to %s\n", ctx->nameOfWire(wire), net->name.c_str(ctx));
+    }
+}
 
-    // Expand from wire specified, either downhill or uphill.
-    //
-    // Expands until it reaches another net of it's own (e.g. source to sink
-    // within site) or a site port (e.g. out to routing network).
-    void expand(WireId wire, const SiteInformation *site_info)
-    {
+struct PossibleSolutions
+{
+    bool tested = false;
+    SiteNetInfo *net = nullptr;
+    std::vector<SitePip>::const_iterator pips_begin;
+    std::vector<SitePip>::const_iterator pips_end;
+};
 
-        bool downhill = site_info->unrouted_source_wires.count(wire) != 0;
-        if (!downhill) {
-            NPNR_ASSERT(site_info->unrouted_sink_wires.count(wire) != 0);
+bool test_solution(SiteArch *ctx, SiteNetInfo *net, std::vector<SitePip>::const_iterator pips_begin,
+                   std::vector<SitePip>::const_iterator pips_end)
+{
+    bool valid = true;
+    std::vector<SitePip>::const_iterator good_pip_end = pips_begin;
+    for (auto iter = pips_begin; iter != pips_end; ++iter) {
+        if (!ctx->bindPip(*iter, net)) {
+            valid = false;
+            break;
         }
 
-        first_wire = wire;
-        net_for_wire = site_info->consumed_wires.at(first_wire);
+        good_pip_end = iter;
+    }
 
-        if (verbose_site_router(ctx)) {
-            log_info("Expanding net %s from %s\n", net_for_wire->name.c_str(ctx), ctx->nameOfWire(first_wire));
+    // Unwind a bad solution
+    if (!valid) {
+        for (auto iter = pips_begin; iter != good_pip_end; ++iter) {
+            ctx->unbindPip(*iter);
         }
+    } else {
+        NPNR_ASSERT(net->driver == ctx->getPipSrcWire(*good_pip_end));
+    }
 
-        completed_routes.clear();
-        wire_to_nodes.clear();
-        node_storage->free_nodes(nodes);
+    return valid;
+}
 
-        auto node = new_node(first_wire, PipId(), /*parent=*/Node());
-        wire_to_nodes[first_wire].push_back(node);
+void remove_solution(SiteArch *ctx, std::vector<SitePip>::const_iterator pips_begin,
+                     std::vector<SitePip>::const_iterator pips_end)
+{
+    for (auto iter = pips_begin; iter != pips_end; ++iter) {
+        ctx->unbindPip(*iter);
+    }
+}
 
-        std::vector<Node> nodes_to_expand;
-        nodes_to_expand.push_back(node);
+bool find_solution_via_backtrack(SiteArch *ctx, std::vector<PossibleSolutions> solutions,
```

Comment:
are `solutions` and `sinks_to_solutions` being passed-by-value rather than const references for a reason?

Reply:
Good catch, let me think through what is going on.  I think this was not causing terrible performance before because this was a tail call, but that is not a reason to leave it like this.

Reply:
Fixed.